### DISCUSSION TITLE
Schema Isolation approach when working with tables used in realtime

### DISCUSSION THREAD

**Discussion by osaxma** [supabase/supabase#5152] [Questions] [answered]
## Problem Statement 
For reference, [Schema Isolation](https://postgrest.org/en/stable/schema_structure.html) as defined in PostgREST doc:
> It is recommended that you don’t expose tables on your API schema. Instead expose views and stored procedures which insulate the internal details from the outside world. This allows you to change the internals of your schema and maintain backwards compatibility. It also keeps your code easier to refactor, and provides a natural way to do API versioning.

Since it's not possible to subscribe to views or even materialized views with realtime (see: https://github.com/supabase/supabase/discussions/3824), are there any known approaches or recommendations to isolate the schema when working with realtime? 

## Use Case

Take a social media app for example where a user could be listening any of the following tables at any given time:
- notifications 
- posts 
- comments 
- messages* 

> \* In fact, `messages` could be a view based on two separate tables (e.g. `message_data` and `message_status`). So the user listen to `message_status` table for rows containing their uid but then they've to fetch the `messages` view after each event to get the full message object. 

If the app listens to these tables directly, then achieving schema isolation for realtime tables becomes difficult. 

## Discussion

I've thought about two approaches so far:
1. **Listen & Fetch**

    The app has to listen to new events that concern them but after each event, they've to fetch the data from the proper view. 
      
    Beside the extra trip, the only other issue is that <s>it's not possible to enforce this on the client when the table in question contains all the necessary data (i.e. AFAIK, realtime does not support column level security).</s> (that was [wrong](https://news.ycombinator.com/item?id=29415302)).

2. **Centralized Events Table** 
    - Create a table for realtime events such as:
    
        | event_id | uid | event_type | api_version* | data (json) |
        | :-: |  :-: |  :-: |  :-: |  :-:  |

    - For each table used in realtime (i.e. notifications, posts, comments, messages), create a trigger to insert the viewable data into the event table (e.g. using `row_to_json` on the corresponding api view of that table). 
    - The user can only listen to the `event` table, only for `INSERT` events and only where uid == their uid using an RLS policy. 
 
    In the suggested event table above, the `api_version` refers to the version of the view. For instance, if there are two views: `api.notifications_view_v1` and `api.notifications_view_v2`, then the trigger should insert two rows -- one per version. The listener, on the other hand, would specify which version they're listening to on the client side. 

    Also the data can be immediately deleted after it was inserted since it'd have no use (i.e. after it has already made it to a `wal` file). 

    By the way, I haven't tried this approach yet and I'm just thinking out loud here.

--- 
Related discussion:
- https://github.com/supabase/supabase/discussions/3424

@steve-chavez -- I would love to hear your input here if you've the time 🙏🏽

**Comment by GaryAustin1** [comment]
I waited on answering here as I can't comment on your schema isolation stuff. 

But in planning for serious realtime updates, you pretty much need to monitor the realtime connection and on any disconnect have a plan for reloading needed table data on the reconnect.   When realtime drops the connection for intermittent internet connection (mainly but not exclusively a mobile device issue) or in the case of Chrome and Edge browser tab going into the background for several minutes, it will attempt to reconnect and you have to monitor the connect status to see that this occurred.  If the connection breaks, you lose any realtime updates during the break in connection, there is no queue.  There is no standard error if you just do .subscribe()... the connection can break and reconnect and there is no error if you don't monitor it.
I do this and then a bunch of stuff on the status change:
```
 .subscribe((status)=>{
                if (status==='SUBSCRIBED') {
                    online.set(true);
                    realtimeLive = true
                }
                else if (status !== 'CLOSED') {
                    realtimeLive = false
                    online.set(false)
                }
         })
```
Based on this I use realtime when it is on the current page (for a chat window for instance) or use it in a status mode (like you mention with a time or record number flag) and then fetch updates on the real table as needed since I have to do this after any connection break anyway.  When a connection reconnects either from internet being lost, or when I put the connection to "sleep" when the tab goes into the background and wakes up, I reload needed info based on a time or record number flag.  Works well for insert only data, but deletes are a pain.

Basically, I believe you need to treat realtime as a "helper" when you need near realtime response, for chat type environments, but the base code needs to refresh all data regularly, or deal with all disconnect/reconnects for maintaining general current data from a table.

**Reply by osaxma** [reply]
Thank you -- these are some useful tips that I haven't thought about thoroughly yet.

**Comment by steve-chavez** [comment]
@osaxma Sorry for the late reply. Note that realtime works independently of the schema, IIRC it only needs a publication on a table, so it doesn't follow the PostgREST's Schema Isolation approach.

So really I think this boils down to "how to work views with realtime", @w3b6x9 @olirice @abc3 might have more input on this.

**Reply by olirice** [reply]
Enabling subscriptions to views would introduce some challenges

Tracking views' table dependencies wouldn't be hard with `pg_depend` but mapping an underpinning `update` to the relevant record(s) in a view is an unsolved problem.

Views also:
- _security risk_ - ignore row level security rules
- _resource risk_ - may be inefficient to query/poll

The same subset of views that are [embeddable](https://postgrest.org/en/latest/api.html#embedding-views) by postgrest and also do not apply row level security might be actionable. We'd have to do some research to see how often that case occurs to gauge the value.

----

At this point the `Listen & Fetch` option is the best bet.

**Comment by osaxma** [comment, marked as answer]
Thank you @steve-chavez and @olirice for your feedback. 

I'm currently experimenting with the `event` table approach. It's roughly looking like this where it supports "listening to view" and avoids "listen & fetch" (i.e. full data is delivered with the event):

```sql

/*----------------  Events TABLE  ---------------------*/


create table public.events ( 
  -- a primary key is necessary for realtime RLS to work 
  id int generated always as identity primary key, 
  -- `null` if the event is public, and the `auth.uid` if the event is for specific user. 
  uid uuid,
  -- customized topic including filters (e.g. "messages_view|{otherUID}" or "comments|{postID}") 
  topic text, 
  -- The inserted/updated data wrapped in a json object
  data json,
  -- `INSERT`, `UPDATE`, or `DELETE` 
  event_type text,
  -- used to delete old events by a cron job
  created_at timestamp with time zone DEFAULT now() NOT NULL
);

/*----------------  REALTIME SETUP  ---------------------*/

-- clients can only listen to the events table and only for insert events. 
ALTER PUBLICATION supabase_realtime SET (publish = 'insert');
ALTER PUBLICATION supabase_realtime ADD TABLE events;


/*----------------- SECURITY  ---------------------*/

ALTER TABLE public.events ENABLE ROW LEVEL SECURITY;

CREATE POLICY events_policy
ON public.events
FOR SELECT USING (
  events.uid is NULL OR events.uid = auth.uid()
);


/*----------------- EXAMPLE FOR "LISTENING" TO A VIEW ---------------------*/
-- In short, create a trigger on the table that would update the view 
-- and the trigger will insert the data from the view into the events table
create or replace function public.create_message_event() 
returns trigger 
as $$
DECLARE 
    _record messages_view;
begin  
  -- get the data from the view
  select * into _record FROM messages_view WHERE uid = NEW.uid and message_id = NEW.message_id;
  -- insert the data into the events table and `realtime` will take it from there
  insert into public.events (uid, topic, event_type, data) 
  values 
    (NEW.uid, CONCAT('messages_view', '|', _record.other_uid), 'INSERT', row_to_json(_record));
  return NEW;
end;
$$ language plpgsql security definer;

create trigger on_new_messages
  after insert on core.messages_status
  for each row execute procedure public.create_message_event();

```


I found this approach useful for:
- Isolating the core schema (i.e. clients cannot listen to any table there).
- "Listening to views" 
- Delivering the full data with the event (eliminates further fetching). 
- Overcoming the current inability to use `eq` more than once in `realtime` filter by customizing the `topic` field to contain all the necessary filters (e.g. on the client side, `supabaseClient.from('events:topic=eq.messages_view|$otherUID').on(...)`. 
- Sending multiple events from a single trigger.
- Allowing more customizability in general.  

There are obvious downsides here such as:
- Added complexity (creating triggers, managing the topic format, cleaning the table, etc.). 
   _Though in my case, I only have four triggers (i.e. new messages, new notifications, new posts, new comments)._
- Unnecessary writes to the events table (and trying to avoid that is another downside).
- Only support simple security rules based on uid (though creating multiple event tables for each type of event would allow more granular security rules without them being complex). 
- ..... I guess I'll figure out more as I go.. 


Any feedback is appreciated.
